### ISSUE TITLE
Separate navigation by module

### ISSUE THREAD

**Issue by bryanl** [vmware-tanzu/octant#340]
The current navigation scheme shows lots of separate data together. Split out 
navigation to show only the current module.

**Comment by GuessWhoSamFoo** [comment]
related #640

**Comment by rossfenrick** [comment]
After several discussions (https://github.com/vmware-tanzu/octant/pull/711) & iterations we came to the conclusion that the vertical nav design https://github.com/vmware-tanzu/octant/issues/640 will not work. From these discussions came the idea to make Octant's information architecture more based around modules (modules being Dashboard, Applications, Cluster, Plugins). The following design is my proposal based on said discussions and feedback.

In the existing navigation of Octant, all of these modules (and their children) are in the left navigation. This results in a lot of information in a small space, three levels of navigation, and presenting distinct information as a single object. By breaking this into separate levels and areas, we can give users a clearer understanding of where they are in the application and better progressive disclosure when they are viewing a single module.

## Module based navigation
![nav-modules-gif](https://user-images.githubusercontent.com/7506366/76449831-fcbaee80-6389-11ea-9ed4-e00a416e6ddd.gif)


This navigation uses [Clarity's Header component](https://clarity.design/documentation/header) with links, separated into 4 modules:
- Dashboard
- Applications
- Cluster
- Plugin (if installed)

### Dashboard
![Dashboard-Overview](https://user-images.githubusercontent.com/7506366/76449927-2d9b2380-638a-11ea-9d7e-86b6a3bff2f9.png)
The default view of the application. This is where users see the objects in their cluster (scoped to a single namespace). Uses Clarity's [Vertical Nav component](https://clarity.design/documentation/vertical-nav/basic-structure/project-pokemon) for left-hand navigation. 

When a left-nav item is selected, the children will be displayed via [tabs](https://clarity.design/documentation/tabs)
![Dashboard-Workloads](https://user-images.githubusercontent.com/7506366/76450248-b023e300-638a-11ea-8632-433db62599c5.png)

When an object is selected, it will navigate to a view, where it's children will also be displayed via [tabs](https://clarity.design/documentation/tabs)
![Dashboard-Individual Workload](https://user-images.githubusercontent.com/7506366/76450360-dba6cd80-638a-11ea-8712-097386130380.png)

This view also introduces back button navigation, so users are not solely dependent on the browser's back button to navigate backwards. A back button is chosen here instead of breadcrumbs, as it is only one level deep.

- Namespace selector and filtering via labels are permanently fixed (always in view) underneath the main header
- Theme selector is moved to the bottom left of the left navigation
- Cluster selector placement and functionality stays as-is

### Applications
![Applications-Overview](https://user-images.githubusercontent.com/7506366/76450761-90d98580-638b-11ea-97c2-061706c4437d.png)
There is no need for a left navigation here, as the content is displayed via cards and can only go one level deep. Namespace & label filter function the same as in Dashboard view.

### Cluster
![Cluster-Overview](https://user-images.githubusercontent.com/7506366/76450864-c1b9ba80-638b-11ea-87ca-f175b68d1e9e.png)
This view is where you see cluster-wide resources. There is no namespace selector, as keeping it would imply that you can scope down to a namespace, which isn't true. Label filter functions the same as elsewhere.

### Plugin
![Plugin-Overview](https://user-images.githubusercontent.com/7506366/76451042-0f362780-638c-11ea-8f05-b941519654ee.png)
If a user has a plugin installed, it will appear at the end of the header as a link, like the other modules. Note that in this navigation model there is no longer a view with a list of plugins with their description and capabilities. I believe this view is redundant with this model, as the list in the header supplies this information. 

Plugins operate in the same way the Dashboard does, where there is a left-hand navigation, a nav item can be selected and it's children will be displayed via tabs, and if an object is selected it will also be displayed via tabs. I use the [ioFog plugin](https://github.com/Edgeworx/octant-iofog-plugin) as an example. Namespace and label filters function the same as elsewhere in the app.

![Pugin-NavItem](https://user-images.githubusercontent.com/7506366/76451468-af8c4c00-638c-11ea-9dd5-1e8711725459.png)
![Pugin-IndividualItem](https://user-images.githubusercontent.com/7506366/76451513-c2068580-638c-11ea-9bd7-9bd48eb00989.png)

What if a user has 6+ plugins installed? My recommendation would be to display as many as possible and then have the last item be a dropdown with the rest of the plugins can be selected. 
![Pugin-NavItem-Lots](https://user-images.githubusercontent.com/7506366/76452012-9f28a100-638d-11ea-9875-16f2938e961f.png)

--------------------
**tldr;** By moving to a module based navigation model it gives the user better a clearer understanding of where they are in the app and where to find objects.

Figma prototype - https://www.figma.com/file/szf5pbvaZAFreHAxb7tM7i/Navigation?node-id=7%3A44127

cc @bryanl @wwitzel3 @GuessWhoSamFoo @mklanjsek @nfarruggiagl

**Comment by mklanjsek** [comment]
This looks very interesting but I am not sure it's a step in the right direction. Need to think some more about it, but here are some initial thoughts. 

1. It seems to me that we are basing our navigation on internal architecture. Why would users care how we internally organized modules? Is that hierarchy really applicable to user perception of their cluster(s)?
2. On most screens we have 3 separate levels of navigation (top nav, left nav and tabs). Some screens have namespace selector, some don't. Applications screen doesn't have left nav at all. If a child link is clicked it will replace its parent with back button and its own set of tabs.  As a result, navigation is all around the place, very inconsistent and potentially very confusing for the end user,
3. I am also not sure about the terminology we use, especially `Dashboard` and `Applications` terms. We could easily call Dashboard what we now call Applications, in fact that is the real dashboard in my view. 

In summary, I would prefer single navigation on the left side where only top level items are visible by default and their children are visible only when selected (like the current nav, just completely collapsed). I would hope that this navigation will be driven by user perception of their cluster instead of internal module organization. Even better, I would prefer to use my mental image of cluster as a backdrop of a graphical navigation element, something like the current Resource Viewer extended to show graphically all objects inside the cluster. Clicking on boxes will open detail screens on the right and detail screens will have only tabs on the top. Hope our internal architecture is flexible enough to handle something like that.

**Comment by rossfenrick** [comment]
> It seems to me that we are basing our navigation on internal architecture. Why would users care how we internally organized modules? Is that hierarchy really applicable to user perception of their cluster(s)?

Navigation is being based off of logical groupings of Octant and a Kubernetes cluster. For example, Namespace scoped objects and cluster scoped objects should not be presented in both the same navigation level and hierarchy as they are now. Separating these uses more of a progressive disclosure model, which improves the focus of the user by reducing clutter and cognitive workload.

> On most screens we have 3 separate levels of navigation (top nav, left nav and tabs). Some screens have namespace selector, some don't. Applications screen doesn't have left nav at all. If a child link is clicked it will replace its parent with back button and its own set of tabs.  As a result, navigation is all around the place, very inconsistent and potentially very confusing for the end user,

I disagree. Different views and sections of an application can have different needs, data to present and use cases to solve. Not every part of an application needs to be represented in the same way, doing so would assume every view is identical. My rationale for application view not having a left nav was "content is displayed via cards and can only go one level deep.". My rationale for no namespace selector was ("There is no namespace selector, as keeping it would imply that you can scope down to a namespace, which isn't true.")

> If a child link is clicked it will replace its parent with back button and its own set of tabs.  As a result, navigation is all around the place, very inconsistent and potentially very confusing for the end user,

This is a very common navigation model. The headings at the top of the page as well as the highlight on the left navigation item give enough affordance to a user of where they are in the application. There are only 2 sets of navigation (left hand nav and tabs), so I disagree it's inconsistent.

> I am also not sure about the terminology we use, especially Dashboard and Applications terms. We could easily call Dashboard what we now call Applications, in fact that is the real dashboard in my view. 

I chose the word dashboard for the k8s objects area because I believe a Dashboard to be a single point for diverse sets of data which allows you to drill down into specific data points/objects to get further insights. In contrast, the Applications view shows one set of data.

**Comment by wwitzel3** [comment]
This is a hard problem and I think we are moving in the right directions with each iteration. As for the naming of things, I'm not super concerned with that, we can defer that until after we get our usage pattern figured out.

I thought about the top nav header some more and played around with a fake mock using clarity and some static HTML and it really does become pretty unmanageable even with just a few plugins. It is a lot of mouse travel and thinking. I don't think it is the optimal solution for us, even though I really do like how it looks. So I agree with Milan here.

To what Milan said about keeping the existing nav, but collapsed by default, and only expanding when selected, I like this idea, but we still have custom resources to deal with.

That is where I get stuck. Do we want to just handled them differently? Nested vertical nav in the content window? I don't know, I'm open to ideas here.

**Comment by rossfenrick** [comment]
If we want to keep the existing navigation but with items collapsed by default and clicking a top-level item both sends you to the root/overview and expands to see it's children, we'll have to make some changes to the Clarity Vertical Nav component, just an FYI.

**Comment by wwitzel3** [comment]
Yeah, it is good to note.

I think it is time for us to just agree that we will not be bound by the limitations of Clarity here and we will make upstream PRs / use other libraries and do what ever we need to, to get the UI the way we want.

**Comment by rossfenrick** [comment]
Here's a .gif of the expandable left-hand navigation prototype a made awhile back for reference
![nav-expanding-selectable](https://user-images.githubusercontent.com/7506366/76554937-f5601780-6453-11ea-8c64-a167169413c8.gif)


Figma URL - https://www.figma.com/proto/szf5pbvaZAFreHAxb7tM7i/Navigation?node-id=1%3A597&viewport=753%2C468%2C0.430318146944046&scaling=min-zoom

**Comment by rossfenrick** [comment]
Pros/Cons of the expandable left-hand nav:
- Pros
  - Scalable. This pattern can hold a lot of items and being able to expand/collapse can help reduce visual clutter and cognitive overload. Probably the biggest advantage of this nav model.
  - Simple. Everything you need to navigate is located in one area.
  - Consistency. This is very similar to the current model.
- Cons
  - If you expand an item it bumps the rest of the navigation out of view. This leads to more scrolling. Also something to note is with Clarity's Vertical Nav component, expanded items stay that way until you explicitly collapse them. So the scrolling issue is compounded the more you navigate.
  - Still have the oddness of heirarchy with the namespace selector and cluster-scoped items. Some items in the nav are affected by the namespace selector, other's aren't. Having the dividers does help this to an extent.

If anyone has ideas for other navigation models let me know. These are the options I've made so far:
Left nav w/ tabs - https://www.figma.com/file/szf5pbvaZAFreHAxb7tM7i/Navigation?node-id=1%3A9
Left nav expandable - https://www.figma.com/file/szf5pbvaZAFreHAxb7tM7i/Navigation?node-id=1%3A10
Header based - https://www.figma.com/file/szf5pbvaZAFreHAxb7tM7i/Navigation?node-id=7%3A44127

If anyone has ideas for tweaks/edits for these existing options I can make some designs around that. I can also make a few design options on how to handle custom resources. Beyond that I'm not sure there's much more exploration I can do on my end.

**Comment by mklanjsek** [comment]
I agree that we should find a way to work around Clarity limitations. I also like the expandable left nav and it shouldn't be hard to make that working one way or another.

**Comment by mklanjsek** [comment]
Thank you @rossfenrick for all models! I will do some prototyping with expandable left nav and see how that would work.

**Comment by mklanjsek** [comment]
Pushed the current to [this branch](https://github.com/nfarruggiagl/octant/tree/640-use-vertical-nav-component-and-design-style) Still some rough edges but it looks promising! I expanded the Clarity Vertical Nav so it now has only one group expanded at the time and the collapsed Nav is showing the group items inside the flyout.

**Comment by wwitzel3** [comment]
I think the goals and ideas for this issue have been generally captured with the merging of #812.
I am going to close this and we can open tickets around the continued implementation of the new navigation, similar to #710
